File: typescript-nodejs/architecture-shell-cqrs/src/behaviors/CachingBehavior.ts
import { BaseRequest } from '../BaseRequest';
import { Query } from '../Query';
import {
  IPipelineBehavior,
  RequestHandlerDelegate,
} from '../IPipelineBehavior';

/**
 * ICache - Cache abstraction for query results
 *
 * Integrate with caching libraries like node-cache, Redis, Memcached, etc.
 */
export interface ICache {
  /**
   * Retrieves cached value if exists
   *
   * @param key - Cache key
   * @returns Cached value or undefined if not found
   */
  get<T>(key: string): Promise<T | undefined>;

  /**
   * Stores value in cache with optional TTL
   *
   * @param key - Cache key
   * @param value - Value to cache
   * @param ttlSeconds - Time-to-live in seconds (optional)
   */
  set<T>(key: string, value: T, ttlSeconds?: number): Promise<void>;

  /**
   * Removes value from cache
   *
   * @param key - Cache key
   */
  remove(key: string): Promise<void>;
}

/**
 * ICacheable - Interface for queries that support caching
 *
 * Implement this interface on queries to enable caching behavior.
 */
export interface ICacheable {
  /**
   * Cache time-to-live in seconds
   */
  readonly cacheTtlSeconds?: number;

  /**
   * Custom cache key (optional)
   * If not provided, key is derived from query properties
   */
  readonly cacheKey?: string;
}

/**
 * CachingBehavior - Pipeline behavior that caches query results
 *
 * Short-circuits handler execution on cache hit (returns cached value).
 * Writes to cache on cache miss.
 * Only applies to queries implementing ICacheable interface.
 *
 * @typeParam TRequest - The request type (must be Query)
 * @typeParam TResponse - The response type
 *
 * @remarks
 * Recommended order: 50 (after telemetry, applies to queries only)
 *
 * Cache key strategy:
 * - If query implements ICacheable with cacheKey, use that
 * - Otherwise, serialize query properties to deterministic JSON string
 * - Include type name to prevent collisions
 *
 * @example
 * ```typescript
 * const behavior = new CachingBehavior(cache);
 * mediator.addBehavior(behavior, new QueryOnlyMatcher());
 * ```
 */
export class CachingBehavior<
  TRequest extends BaseRequest<TResponse>,
  TResponse
> implements IPipelineBehavior<TRequest, TResponse>
{
  readonly order = 50;

  constructor(private readonly cache: ICache) {}

  async handle(
    request: TRequest,
    next: RequestHandlerDelegate<TResponse>,
    _signal: AbortSignal
  ): Promise<TResponse> {
    // Only cache queries that implement ICacheable
    if (!this.isCacheable(request)) {
      return await next();
    }

    // Generate cache key
    const cacheKey = this.generateCacheKey(request);

    // Check cache
    const cachedValue = await this.cache.get<TResponse>(cacheKey);
    if (cachedValue !== undefined) {
      return cachedValue;
    }

    // Cache miss - execute handler
    const response = await next();

    // Write to cache
    const ttl = (request as ICacheable).cacheTtlSeconds;
    await this.cache.set(cacheKey, response, ttl);

    return response;
  }

  /**
   * Type guard to detect cacheable queries
   */
  private isCacheable(request: any): request is Query<any> & ICacheable {
    return this.isQuery(request);
    // Note: Could add additional check for ICacheable interface
    // return this.isQuery(request) && 'cacheTtlSeconds' in request;
  }

  /**
   * Type guard to detect query requests
   */
  private isQuery(request: any): request is Query<any> {
    return (
      '_isQuery' in request ||
      request.constructor.name.endsWith('Query') ||
      Object.getPrototypeOf(request).constructor.name.endsWith('Query')
    );
  }

  /**
   * Generates deterministic cache key from query properties
   */
  private generateCacheKey(request: TRequest & ICacheable): string {
    // Use custom cache key if provided
    if (request.cacheKey) {
      return `${request.constructor.name}:${request.cacheKey}`;
    }

    // Serialize query properties to JSON with sorted keys
    const properties = { ...request };
    delete (properties as any).cacheTtlSeconds;
    delete (properties as any).cacheKey;

    const serialized = JSON.stringify(properties, Object.keys(properties).sort());
    return `${request.constructor.name}:${serialized}`;
  }
}

File: python-django/architecture-shell-cqrs/src/architecture_shell_cqrs/concrete_behaviors/caching_behavior.py
"""CachingBehavior - Query result caching."""

import json
from typing import Protocol, TypeVar, Optional, Any
from architecture_core.functional import Result
from architecture_shell_cqrs.behaviors import IPipelineBehavior
from architecture_shell_cqrs.requests import IQuery

TRequest = TypeVar("TRequest")
TResponse = TypeVar("TResponse")


class ICacheProvider(Protocol):
    """
    Cache abstraction for query results.

    Integrate with caching libraries like Redis, Memcached, Django cache, etc.
    """

    async def get(self, key: str) -> Optional[Any]:
        """
        Retrieve cached value if exists.

        Args:
            key: Cache key

        Returns:
            Cached value or None if not found
        """
        ...

    async def set(self, key: str, value: Any, ttl_seconds: Optional[int] = None) -> None:
        """
        Store value in cache with optional TTL.

        Args:
            key: Cache key
            value: Value to cache
            ttl_seconds: Time-to-live in seconds (optional)
        """
        ...

    async def remove(self, key: str) -> None:
        """
        Remove value from cache.

        Args:
            key: Cache key
        """
        ...


class ICacheable(Protocol):
    """
    Interface for queries that support caching.

    Implement this protocol on queries to enable caching behavior.
    """

    @property
    def cache_ttl_seconds(self) -> Optional[int]:
        """Cache time-to-live in seconds."""
        ...

    @property
    def cache_key(self) -> Optional[str]:
        """
        Custom cache key (optional).

        If not provided, key is derived from query properties.
        """
        ...


class CachingBehavior(IPipelineBehavior[TRequest, TResponse]):
    """
    Pipeline behavior that caches query results.

    Short-circuits handler execution on cache hit (returns cached value).
    Writes to cache on cache miss.
    Only applies to queries implementing ICacheable interface.

    **Recommended order:** 50 (after telemetry, applies to queries only)

    **Cache key strategy:**
    - If query implements ICacheable with cache_key, use that
    - Otherwise, serialize query properties to deterministic JSON string
    - Include type name to prevent collisions

    Example:
        ```python
        behavior = CachingBehavior(cache_provider)
        mediator.register_behavior(behavior)
        ```
    """

    order = 50

    def __init__(self, cache: ICacheProvider):
        """
        Initialize CachingBehavior.

        Args:
            cache: Cache provider implementation
        """
        self.cache = cache

    async def handle(self, request: TRequest, next_handler) -> Result[TResponse]:
        """
        Handle request with caching.

        Args:
            request: The request being processed
            next_handler: Delegate to invoke next behavior or handler

        Returns:
            Result containing response or error
        """
        # Only cache queries that implement ICacheable
        if not self._is_cacheable(request):
            return await next_handler()

        # Generate cache key
        cache_key = self._generate_cache_key(request)

        # Check cache
        cached_value = await self.cache.get(cache_key)
        if cached_value is not None:
            return cached_value

        # Cache miss - execute handler
        response = await next_handler()

        # Write to cache
        ttl = getattr(request, "cache_ttl_seconds", None)
        await self.cache.set(cache_key, response, ttl)

        return response

    def _is_cacheable(self, request) -> bool:
        """Type guard to detect cacheable queries."""
        return self._is_query(request)

    def _is_query(self, request) -> bool:
        """Type guard to detect query requests."""
        return isinstance(request, IQuery)

    def _generate_cache_key(self, request) -> str:
        """Generate deterministic cache key from query properties."""
        # Use custom cache key if provided
        custom_key = getattr(request, "cache_key", None)
        if custom_key:
            return f"{type(request).__name__}:{custom_key}"

        # Serialize query properties to JSON with sorted keys
        # Exclude caching metadata from the cache key calculation
        properties = {
            k: v
            for k, v in vars(request).items()
            if k not in ("cache_ttl_seconds", "cache_key")
        }

        serialized = json.dumps(properties, sort_keys=True)
        return f"{type(request).__name__}:{serialized}"


__all__ = ["CachingBehavior", "ICacheProvider", "ICacheable"]
